peer/x/circus/list.go

```diff
@@ -0,0 +1,368 @@
+
+	// If we deplete the entire low concurrency ring, we swap the high
+	// concurrency ring.
+	// Assuming relatively homogeneous latency for all requests,
+	// this would usually occur only when concurrency for the whole list is
+	// climbing.
+	// When latency is not homogeneous, well, perfect load distribution is the
+	// enemy of fast load distribution.
+	if l.nodes.empty(l.lo) {
+		l.hi, l.lo = l.lo, l.hi
+	}
+
+	peer := l.peers[index]
+	onFinish := l.subscribers[index].boundOnFinish
+	return peer, onFinish
+}
+
+func (l *List) onFinish(subscriber *subscriber) {
+	l.mx.Lock()
+	defer l.mx.Unlock()
+
+	// There is some chance that the onFinish hook will be called long after
+	// the underlying node has been moved to another list.
+	// This would cause a problem if the peer has been freed, so we abort
+	// early in that case.
+	// We also abort early if the peer is no longer connected so we don't
+	// accidentally put it back into rotation.
+	index := subscriber.index
+	if p := l.peers[index]; p == nil || p.Status().ConnectionStatus == peer.Unavailable {
+		return
+	}
+	// Event accounting for the above cases, there remains a chance
+	// that the peer was released and a new peer was retained, reusing this subscriber.
+	// However, the only consequence of proceeding here is being selected later
+	// and this should be rare enough as not to cause any material trouble.
+	// Certainly less trouble that allocating a unique subscriber for each retained
+	// peer and keeping it in a map for the later release.
+	l.nodes.shift(index, l.lo)
+}
+
+// Update receives changes to the peer list's membership from a source like
+// DNS.
+func (l *List) Update(updates peer.ListUpdates) error {
+	l.mx.Lock()
+	defer l.mx.Unlock()
+
+	var errs error
+
+	for _, pid := range updates.Removals {
+		addr := pid.Identifier()
+		index := l.index[addr]
+		// index, exists := l.index[addr]
+		// if !exists {
+		// 	panic(fmt.Sprintf("remove never added: %s", addr))
+		// }
+		delete(l.index, addr)
+		if index > 0 {
+			// if l.peers[index].Identifier() != addr {
+			// 	panic(fmt.Sprintf("expected peer identifiers to match: %d %s %s", index, l.peers[index].Identifier(), addr))
+			// }
+			// remove from index of known peers.
+			// release the peer for garbage collection to avoid heap
+			// bloat.
+			l.peers[index] = nil
+			l.statuses[index] = peer.Unavailable
+			// move the node to the end of the free list for later
+			// reuse.
+			l.nodes.shift(uint8(index), _free)
+			err := l.transport.ReleasePeer(pid, &l.subscribers[index])
+			if err != nil {
+				errs = multierr.Append(errs, err)
+			}
+		}
+	}
+
+	additions := updates.Additions
+	if !l.options.noShuffle {
+		additions = shuffle(l.randSrc, additions)
+	}
+
+	for _, pid := range additions {
+		addr := pid.Identifier()
+		// if _, exists := l.index[addr]; exists {
+		// 	panic(fmt.Sprintf("add already added: %s", addr))
+		// }
+		l.index[addr] = 0
+		// The maximum capacity of this peer list is 252 peers.
```

Comment:
Why do we have this cap? This should be called out in the top-level package docs.